Handle io errors gracefully

konstin opened the pull request:
## Summary

It can happen that we can't read a file (a python file, a jupyter notebook or pyproject.toml), which needs to be handled and handled consistently for all file types. Instead of using `Err` or `error!`, we emit E602 with the io error as message and continue. This PR makes sure we handle all three cases consistently, emit E602.

I'm not convinced that it should be possible to disable io errors, but we now handle the regular case consistently and at least print warning consistently.

I went with `warn!` but i can change them all to `error!`, too.

It also checks the error case when a pyproject.toml is not readable. The error message is not very helpful, but it's now a bit clearer that actually ruff itself failed instead vs this being a diagnostic.

## Examples

This is how an Err of `run` looks now:

![image](https://github.com/astral-sh/ruff/assets/6826232/890f7ab2-2309-4b6f-a4b3-67161947cc83)

With an unreadable file and `IOError` disabled:

![image](https://github.com/astral-sh/ruff/assets/6826232/fd3d6959-fa23-4ddf-b2e5-8d6022df54b1)

(we lint zero files but count files before linting not during so we exit 0)

I'm not sure if it should (or if we should take a different path with manual ExitStatus), but this currently also triggers when `files` is empty:

![image](https://github.com/astral-sh/ruff/assets/6826232/f7ede301-41b5-4743-97fd-49149f750337)

## Test Plan

Unix only: Create a temporary directory with files with permissions `000` (not readable by the owner) and run on that directory. Since this breaks the assumptions of most of the test code (single file, `ruff` instead of `ruff_cli`), the test code is rather cumbersome and looks a bit misplaced; i'm happy about suggestions to fit it in closer with the other tests or streamline it in other ways. I added another test for when the entire directory is not readable.

konstin commented on crates/ruff_cli/tests/integration_test.rs:
```diff
@@ -278,3 +291,29 @@ Found 1 error.
+        .mode(0o000)
+        .open(pyproject_toml)?;
+
+    // Don't `--isolated` since the configuration discovery is where the error happens
```
Is configuration discovery the right term?

charliermarsh replied:
Yes!

charliermarsh commented on crates/ruff/src/pyproject_toml.rs:
```diff
@@ -22,34 +21,33 @@ struct PyProjectToml {
+                    messages
+                } else {
+                    Vec::new()
+                };
```
Can't we not just push inside the `if`, and then return `messages` either way -- since it'd be empty in the `else` case, right?

charliermarsh commented on crates/ruff_cli/src/commands/run.rs:
```diff
@@ -226,3 +226,82 @@ with the relevant file contents, the `pyproject.toml` settings, and the followin
 }
+
+#[cfg(test)]
+#[cfg(unix)] // Remove when adding a second test
```
Should this be `TODO(konsti)`?

konstin replied:
no it's just a note for when someone wants to add another test case so they don't trip over this

charliermarsh commented on crates/ruff/src/settings/mod.rs:
```diff
@@ -293,15 +293,13 @@ impl Settings {
         }
     }
 
-    #[cfg(test)]
```
Why are these no longer `#[cfg(test)]`? Are they used outside of tests?

charliermarsh replied:
Oh, perhaps for the configuration tests.

konstin replied:
they are used across crates now (the test has to live in ruff_cli)) and `cfg(test)` doesn't get passed across crate boundaries

charliermarsh commented on crates/ruff_cli/src/diagnostics.rs:
```diff
@@ -135,9 +151,14 @@ pub(crate) fn lint_path(
+                }
+            };
             let source_file = SourceFileBuilder::new(path.to_string_lossy(), contents).finish();
-            lint_pyproject_toml(source_file, &settings.lib)?
```
Are these not already treated as `IOError` via the `IOError` handling in `run.rs`? Like this block:

```rust
error!(
    "{}{}{} {message}",
    "Failed to lint ".bold(),
    fs::relativize_path(path).bold(),
    ":".bold()
);
let settings = resolver.resolve(path, pyproject_config);
if settings.rules.enabled(Rule::IOError) {
    let file =
        SourceFileBuilder::new(path.to_string_lossy().as_ref(), "").finish();

    Diagnostics::new(
        vec![Message::from_diagnostic(
            Diagnostic::new(IOError { message }, TextRange::default()),
            file,
            TextSize::default(),
        )],
        ImportMap::default(),
    )
} else {
    Diagnostics::default()
}
```

charliermarsh replied:
With this change, do we not lose that warning message that appears separately from the diagnostics?

charliermarsh commented on crates/ruff_cli/src/bin/ruff.rs:
```diff
@@ -50,7 +50,15 @@ pub fn main() -> ExitCode {
+                // "failed to read file: subdir/pyproject.toml")
+                for cause in err.chain() {
+                    eprintln!("  {} {cause}", "Caused by:".bold());
+                }
```
Do you mind including a screenshot of this in the PR summary? Trying to visualize it.

konstin replied:
done

konstin commented:
the diagnostics now check whether `IOError` is enabled and if it isn't, print a `warn!` instead (i can change this to an `error!`, not sure which one is better here). I added another test case for an unreadable directory. There are still cases where we didn't lint any file and still exit 0 because `IOError` is disabled and/or we count paths before linting instead of after. I added screenshots to the original change description

charliermarsh commented on crates/ruff_cli/src/commands/run.rs:
```diff
@@ -226,3 +226,85 @@ with the relevant file contents, the `pyproject.toml` settings, and the followin
 }
+
+#[cfg(test)]
+#[cfg(unix)] // If you need to add a second test, move this cfg
```
I might just suggest removing this comment -- it was confusing enough for me that I had to ask about it :joy:

konstin replied:
fair

charliermarsh commented on crates/ruff_cli/src/bin/ruff.rs:
```diff
@@ -50,7 +50,15 @@ pub fn main() -> ExitCode {
-                eprintln!("{}{} {err:?}", "error".red().bold(), ":".bold());
+                // This communicates that this isn't a linter error but ruff itself hard-errored for
+                // some reason (e.g. failed to resolve the configuration)
+                eprintln!("{}", "ruff failed".red().bold());
```
Maybe, for stylistic consistency with the warnings, we do:

```
<bold red>error</><bold>: ruff failed</bold>
  Cause ...
  Cause ...

konstin replied:
done

charliermarsh commented on crates/ruff_cli/Cargo.toml:
```diff
@@ -64,6 +64,8 @@ wild = { version = "2" }
 
 [dev-dependencies]
 assert_cmd = { version = "2.0.8" }
+insta = { workspace = true, features = ["filters"] }
```
Do we need to specify features if they're specified at the workspace level? (I genuinely don't know.)

konstin replied:
surprisingly, yes
